### PR TITLE
HGCAL Pattern Recognition using FastJet

### PR DESCRIPTION
**Pull request by rovere**
#### PR description:

A novel pattern recognition plugin has been added to the `TICL` framework based on `FastJet`.
At present, only `anti-kt` clustering algorithm has been added. All algorithms from the rich `FastJet` library could be tested using different configurations.
A `processModifier` has been added to allow the end-user to enable this plugin. The modifier is called `fastJetTICL`.
The modifier has been added in such a way that all the validation plots for the Tracksters produced by the `FastJet` pattern recognition will be automatically produced.
At present jets are translated into `Tracksters`. As a future development, if needed, we could save also the reconstructed `jets`.
It would be nice to have this code integrated into the release, even if not bundled to any official workflow, to allow interested people to develop on top of that.

You can find some more information about its `performance` (it's quite hard to call them as such, as of yet...) [here](https://indico.cern.ch/event/1088389/contributions/4575464/attachments/2331927/3974172/20211021_HGCAL_WM_FastJet_MR.pdf).
This PR has also been recently _announced_ during [this](https://indico.cern.ch/event/1095576/contributions/4608272/attachments/2345071/3998685/HGCAL%20Reco%20meeting%20November%202021.pdf) reconstruction meeting.

#### PR validation:

Run on `Phase2` dedicated workflows and private _multiparticle_ samples w/o issues.

### CODE REVIEW

**Comment by rovere** [comment]
assign hgcal-dpg

**Comment by rovere** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
For my understanding, is the TF model the same across there four modules:
- https://github.com/cms-sw/cmssw/blob/c3022297c9725a1fde784f0c13c1b48139877c8c/RecoHGCal/TICL/plugins/PatternRecognitionbyFastJet.cc#L219
- https://github.com/cms-sw/cmssw/blob/c3022297c9725a1fde784f0c13c1b48139877c8c/RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc#L594
- https://github.com/cms-sw/cmssw/blob/c3022297c9725a1fde784f0c13c1b48139877c8c/RecoHGCal/TICL/plugins/PatternRecognitionbyCA.cc#L384
- https://github.com/cms-sw/cmssw/blob/c3022297c9725a1fde784f0c13c1b48139877c8c/RecoHGCal/TICL/plugins/PatternRecognitionbyCLUE3D.cc#L354

Perhaps there is an opportunity to reduce code duplication here by a shared evaluator.

**Comment by rovere** [comment]
Ciao @jpata yes indeed, the model is the same, for the time being.
In parallel, we are working on the factorization that was already suggested some months ago to save memory by re-using the same session for the same model in several plugins. That will come, at some point in the near future.
In a little farther away future, depending on the use cases for the different plugins, it is quite likely we will have different models for the different plugins. In this sense, the shared evaluator will be less important, since shaping batching and everything that comes with that will likely be different.
Benedikt, as you know, is actively working on this front.

**Comment by jpata** [comment]
Perhaps a workflow can be introduced if this is intended for production:
- https://github.com/cms-sw/cmssw/blob/master/Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py
- https://github.com/cms-sw/cmssw/blob/master/Configuration/PyReleaseValidation/python/relval_2026.py

**Comment by rovere** [comment]
Ciao @jpata I took the same path we had taken for the `CLUE3D` case, i.e., add a `procModifier` w/o any additional proliferation of dedicated workflows.
If/when at some point this PR algorithm will be part of the Phase2 reconstruction, it will automatically be monitored. Up until now, I believe that the `procModifier` is small and practical enough to allow users to investigate the new possibilities brought in by `FastJet`.
What do you think?

**Comment by jpata** [comment]
Generally we try to make an effort that new functionality would also be regularly tested in IBs, when possible and practical.
How many different configurations of modifiers are there for HGCAL reconstruction? Perhaps it could be a good time to introduce some workflows for the major ones.
Also, do you have some pre/post numbers on CPU and memory consumption with igprof?

**Comment by rovere** [comment]
Ciao @jpata 
there are currently 2 `procModifiers`, one for `CLUE3D` and another one for  `FastJet`. Well, perhaps I should have said one, since this PR is not merged.
I have numbers in the slides linked in the description. Definitely not conclusive and for sure nothing `igprof` related.
Yet a good comparison against what is currently run by default in release.
What are you practically suggesting to do in order to integrate this PR?

**Comment by jpata** [comment]
- a test workflow
- igprof CPU and MEM numbers

**Comment by jpata** [comment]
An alternative to a workflow is that you enable the modifier here temporarily, we run the tests and get some numbers, then you disable the modifier again.

**Comment by jpata** [comment]
kind ping on this @rovere, this PR can move forward fairly easily I think
> enable the modifier here temporarily, we run the tests and get some numbers, then disable the modifier again.

**Comment by jpata** [comment]
Just in case, this may be helpful to enable the new functionality for the bot test and profiling. Experts may suggest a cleaner way, but it works AFAICT.

```diff
diff --git a/Configuration/Eras/python/Era_Phase2_cff.py b/Configuration/Eras/python/Era_Phase2_cff.py
index 51e9a8d63cb..5ab11b71cda 100644
--- a/Configuration/Eras/python/Era_Phase2_cff.py
+++ b/Configuration/Eras/python/Era_Phase2_cff.py
@@ -7,6 +7,7 @@ from Configuration.Eras.Modifier_phase2_ecal_cff import phase2_ecal
 from Configuration.Eras.Modifier_trackingPhase2PU140_cff import trackingPhase2PU140
 from Configuration.Eras.Modifier_phase2_hcal_cff import phase2_hcal
 from Configuration.Eras.Modifier_phase2_hgcal_cff import phase2_hgcal
+from Configuration.ProcessModifiers.fastJetTICL_cff import fastJetTICL
 from Configuration.Eras.Modifier_phase2_muon_cff import phase2_muon
 from Configuration.Eras.Modifier_phase2_GEM_cff import phase2_GEM
 from Configuration.Eras.Modifier_phase1Pixel_cff import phase1Pixel
@@ -18,4 +19,4 @@ from Configuration.Eras.Modifier_phase2_trigger_cff import phase2_trigger
 from Configuration.Eras.Modifier_ctpps_2021_cff import ctpps_2021
 from Configuration.ProcessModifiers.dd4hep_cff import dd4hep
 
-Phase2 = cms.ModifierChain(Run3_noMkFit.copyAndExclude([phase1Pixel,trackingPhase1,ctpps_2021,dd4hep]), phase2_common, phase2_tracker, trackingPhase2PU140, phase2_ecal, phase2_hcal, phase2_hgcal, phase2_muon, phase2_GEM, hcalHardcodeConditions, phase2_timing, phase2_timing_layer, phase2_trigger)
+Phase2 = cms.ModifierChain(Run3_noMkFit.copyAndExclude([phase1Pixel,trackingPhase1,ctpps_2021,dd4hep]), phase2_common, phase2_tracker, trackingPhase2PU140, phase2_ecal, phase2_hcal, phase2_hgcal, phase2_muon, phase2_GEM, hcalHardcodeConditions, phase2_timing, phase2_timing_layer, phase2_trigger, fastJetTICL)
```

**Comment by rovere** [comment]
Dear @jpata 
I'll try to work on this PR this week to have it integrated soon.
Thanks.
Marco.

**Comment by rovere** [comment]
Ciao @jpata 
I ended up adding dedicated workflows for both `CLUE3D` and `FastJet`.
I also profiled the `FastJet` on a `TTbar+PU` sample with `igprof` with the performance profiler. The bottom line is that `FastJet`, as it is configured right now, is much faster than the `CA` (well, that's not correct, since the `CA` is run 4 times and they all look `C++-the-same-type` to `igprof`).
If you could suggest an effective way of sharing the `igprof` profile, I'd be happy to do that.

Let me know if this is an acceptable solution.

**Comment by jpata** [comment]
test parameters:
  - workflows = 34834.201, 34834.202

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
@rovere thanks for the update. Note that the bot can only profile the "default" Phase2 workflow 23434.21.

Can you please post some igprof CPU and MEM profiles for this PR, along with event size changes, comparing 34834.201, 34834.202 to some baseline?
- https://twiki.cern.ch/twiki/bin/viewauth/CMS/RecoIntegration#Run_profiler_igprof
- https://twiki.cern.ch/twiki/bin/viewauth/CMS/RecoIntegration#Compare_output_products

**Comment by rovere** [comment]
Ciao @jpata 
I run the performance profiler (not the memory one) using the newly introduced workflow 35234.202. `201` is, in some sense, out of the scope of this PR.
If you prefer to have profiles for the workflows you have suggested, it will take another day or running. Let me know what you prefer.

In the meantime, the profile I produced is available at [this link](https://rovere.web.cern.ch/rovere/ig.FastJetTICL_PU.9.sql3).

**Comment by rovere** [comment]
`34834.201, 34834.202` are not among the workflows I had enabled.

**Comment by rovere** [comment]
test parameters:

  * workflows = 35034.201,35034.202

**Comment by rovere** [comment]
@cmsbuild please test

**Comment by rovere** [comment]
@jpata any idea why the `bot` cannot find the workflows?
In my local test area, I was able to run and profile them. Is there anything I somehow missed to include in the PR?

**Comment by rovere** [comment]
@cmsbuild please test

**Comment by rovere** [comment]
test parameters:
   * workflows = 35034.201,35034.202
   * relvals_opt= -w upgrade

**Comment by rovere** [comment]
@cmsbuild  please test

**Comment by jpata** [comment]
I uploaded the profile you sent here: http://cms-reco-profiling.web.cern.ch/cms-reco-profiling/cgi-bin/igprof-navigator/prs/36172/ig.FastJetTICL_PU.9/296 

Screenshot 2021-12-14 at 13 57 36

Can we conclude that the fastjet-based approach is faster by a factor ~3-4x, or is this not an apples to apples comparison? In any case, I think it's fine in terms of the reco CPU performance checks.
It would be good to check and report memory and event size as well, just in case.

Concerning the workflows, I checked out your code locally and 34834.201, 34834.202 seemed to be there in the matrix according to `runTheMatrix.py -w upgrade -n`.

```
Traceback (most recent call last):
  File "/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/36172/21248/CMSSW_12_3_X_2021-12-13-1100/bin/slc7_amd64_gcc900/runTheMatrix.py", line 598, in <module>
    ret = runSelected(opt)
  File "/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/36172/21248/CMSSW_12_3_X_2021-12-13-1100/bin/slc7_amd64_gcc900/runTheMatrix.py", line 31, in runSelected
    if len(undefSet)>0: raise ValueError('Undefined workflows: '+', '.join(map(str,list(undefSet))))
ValueError: Undefined workflows: 35034.202, 35034.201
```
I'm not sure about this error.

**Comment by rovere** [comment]
Thanks @jpata 
I think the `bot` was confused because it was not looking for the workflows in the `upgrade` section. I tried to tell it to do so: let's see if I got the syntax ok.

I'd conclude that `FastJet` is as fast as a single iteration of the `CA`, given the fact that we run 4 of them and they have the same `C++` type. 
I'm rather satisfied to see no time explosion, at this stage.

I can work on the memory report later today/tomorrow.

In the meantime let's see if the `bot` is happy, this time.

**Comment by rovere** [comment]
test parameters:
   * workflows = 35034.201,35034.202
   * relvals_opt= -w all

**Comment by rovere** [comment]
@cmsbuild please test

**Comment by rovere** [comment]
test parameters:
   * workflows = 35034.201,35034.202
   * relvals_opt= -w standard,highstats,pileup,generator,extendedgen,production,upgrade,cleanedupgrade,ged

**Comment by rovere** [comment]
@cmsbuild please test

**Comment by rovere** [comment]
Ciao @jpata 
does that ring a bell??

```
ERROR reading file: relval_upgrade 'HARVESTFakeHLT_trackingMkFit_2017'
15:29:15 Traceback (most recent call last):
15:29:15   File "/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/36172/21257/CMSSW_12_3_X_2021-12-13-2300/bin/slc7_amd64_gcc900/runTheMatrix.py", line 598, in <module>
15:29:15     ret = runSelected(opt)
15:29:15   File "/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/36172/21257/CMSSW_12_3_X_2021-12-13-2300/bin/slc7_amd64_gcc900/runTheMatrix.py", line 23, in runSelected
15:29:15     mrd.prepare(opt.useInput, opt.refRel, opt.fromScratch)
15:29:15   File "/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/36172/21257/CMSSW_12_3_X_2021-12-13-2300/python/Configuration/PyReleaseValidation/MatrixReader.py", line 506, in prepare
15:29:15     self.readMatrix(matrixFile, useInput, refRel, fromScratch)
15:29:15   File "/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/36172/21257/CMSSW_12_3_X_2021-12-13-2300/python/Configuration/PyReleaseValidation/MatrixReader.py", line 251, in readMatrix
15:29:15     if self.relvalModule.steps[stepName] is None:
15:29:15 KeyError: 'HARVESTFakeHLT_trackingMkFit_2017'
```

**Comment by perrotta** [comment]
@rovere see #36477 for the possible fix (I imagine...)

**Comment by rovere** [comment]
Ciao @perrotta thanks for that.
You that talk `cms-bot` language way better than me, could you suggest how to have this PR tested with #36477 with the changes I included above?

**Comment by rovere** [comment]
@cmsbuild please test with #36477

**Comment by perrotta** [comment]
> @cmsbuild please test with #36477

Indeed!  :-)
I didn't start them because I was not sure if you preferred to wait for the results of the previos tests anyhow...

**Comment by rovere** [comment]
Dear @cmsbuild 
I'm quite fed up with your behaviour.
I'll leave the handling of the testing of this PR to someone else 😄

**Comment by rovere** [comment]
In the meantime, you can find the `MEM_LIVE` profile [here](https://rovere.web.cern.ch/rovere/ig.FastJetTICL_PU_MEM_LIVE.9.sql3).
Again, really nothing to report from my side.

**Comment by jpata** [comment]
Thanks for the profiling report and for the patience with the PR tests!

http://cms-reco-profiling.web.cern.ch/cms-reco-profiling/cgi-bin/igprof-navigator/prs/36172/ig.FastJetTICL_PU_MEM_LIVE.9/659
Screenshot from 2021-12-15 16-07-59

Looks like 10MB used by FastJet in the tracksters loop, so we confirm it's fine for reco.

**Comment by jpata** [comment]
The latest available IB is still broken as far as I can tell (unrelated reason to this PR, arising from https://github.com/cms-sw/cmssw/pull/36167 AFAIK)
```
  File "/home/joosep/reco/36172/CMSSW_12_3_X_2021-12-14-2300/python/Configuration/PyReleaseValidation/MatrixReader.py", line 251, in readMatrix
    if self.relvalModule.steps[stepName] is None:
KeyError: 'HARVESTFakeHLT_trackingMkFit_2017'
```
so we might need to wait until tomorrow to run the test.

The following seems to be working, so I don't have doubts about this PR:
```
joosep@joosep-desktop-work:~/reco/36172/CMSSW_12_3_0_pre1$ runTheMatrix.py -w upgrade -l 35034.201,35034.202
...
Preparing to run 35034.201 TTbar_14TeV+2026D77_ticl_clue3D+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal

# in: /home/joosep/reco/36172/CMSSW_12_3_0_pre1 going to execute cd 35034.201_TTbar_14TeV+2026D77_ticl_clue3D+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal
 cmsDriver.py TTbar_14TeV_TuneCP5_cfi  -s GEN,SIM -n 10 --conditions auto:phase2_realistic_T21 --beamspot HLLHC14TeV --datatier GEN-SIM --eventcontent FEVTDEBUG --geometry Extended2026D77 --era Phase2C11I13M9 --relval 9000,100 --fileout file:step1.root  > step1_TTbar_14TeV+2026D77_ticl_clue3D+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal.log  2>&1
 

Preparing to run 35034.202 TTbar_14TeV+2026D77_ticl_FastJet+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal

# in: /home/joosep/reco/36172/CMSSW_12_3_0_pre1 going to execute cd 35034.202_TTbar_14TeV+2026D77_ticl_FastJet+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal
 cmsDriver.py TTbar_14TeV_TuneCP5_cfi  -s GEN,SIM -n 10 --conditions auto:phase2_realistic_T21 --beamspot HLLHC14TeV --datatier GEN-SIM --eventcontent FEVTDEBUG --geometry Extended2026D77 --era Phase2C11I13M9 --relval 9000,100 --fileout file:step1.root  > step1_TTbar_14TeV+2026D77_ticl_FastJet+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal.log  2>&1
```

**Comment by rovere** [comment]
Ciao @jpata 
ok, so I won't trigger any more tests and wait for a proper IB.

Thanks,
M.

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
test parameters:
  - workflows = 35034.201,35034.202
  - relvals_opt= -w upgrade

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
I'm confused by the last bot error
```
Validation/Performance/TimeMemoryJobReport.customiseWithTimeMemoryJobReport'], ['DoubleMuFlatPt2To100Dxy100GunProducer_2026D88PU_GenSimHLBeamSpot_PMXS1S2PR', 'PREMIX_2026D88PU_PremixHLBeamSpotPU_PMXS1S2PR', 'DigiTriggerPU_PMXS1S2PR_2026D88PU', 'RecoGlobalPU_PMXS1S2PR_2026D88PU', 'HARVESTGlobalPU_PMXS1S2PR_2026D88PU'])
ignoring non-requested file relval_gpu
ignoring non-requested file relval_2017
ignoring non-requested file relval_2026
ignoring non-requested file relval_identity
ignoring non-requested file relval_machine
ignoring non-requested file relval_premix
Traceback (most recent call last):
  File "/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/36172/21312/CMSSW_12_3_X_2021-12-15-2300/bin/slc7_amd64_gcc10/runTheMatrix.py", line 598, in <module>
    ret = runSelected(opt)
  File "/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/36172/21312/CMSSW_12_3_X_2021-12-15-2300/bin/slc7_amd64_gcc10/runTheMatrix.py", line 31, in runSelected
    if len(undefSet)>0: raise ValueError('Undefined workflows: '+', '.join(map(str,list(undefSet))))
ValueError: Undefined workflows: 250202.181
```

maybe it's because of the many additional relval opts? I'm not sure.

**Comment by jpata** [comment]
@cms-sw/upgrade-l2 sorry to bother, do you know what might be going on with the matrix here? do we need some special options for the bot?

**Comment by AdrianoDee** [comment]
As far as I remember (don't know why) apparently extra workflows cannot be inputed to the bot, if one want to be tested they should be placed in `relval_2026 `. Discovering it on my skin for #36235.

**Comment by jpata** [comment]
test parameters:

**Comment by rovere** [comment]
+1

**Comment by jpata** [comment]
+reconstruction
- new FastJet modules for TICL
- new workflows to test the code: 35034.201,35034.202 (these were run offline and not in the bot)
- CPU profiling here: https://github.com/cms-sw/cmssw/pull/36172#issuecomment-993471014
- memory profiling here: https://github.com/cms-sw/cmssw/pull/36172#issuecomment-994826210

**Comment by AdrianoDee** [comment]
+upgrade

**Comment by jpata** [comment]
kind ping @cms-sw/pdmv-l2

**Comment by bbilin** [comment]
+1

**Comment by rovere** [comment]
Hello there,
another kind ping @cms-sw/pdmv-l2

**Comment by bbilin** [comment]
sorry for the latency, just got back. 
+1

**Comment by rovere** [comment]
thanks!

**Comment by bbilin** [comment]
+pdmv

**Comment by qliphy** [comment]
please test
just to refresh the test results

**Comment by perrotta** [comment]
+operations
- Two new special TICL pattern recognition workflows defined for clue3D (not related to this PR, but welcome anyhow) and FastJet
- They have been succesfully tested offline, and should become available to bot once this PR will be merged

File: RecoHGCal/TICL/plugins/PatternRecognitionbyFastJet.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    void makeTracksters(const typename PatternRecognitionAlgoBaseT<TILES>::Inputs& input,
+                        std::vector<Trackster>& result,
+                        std::unordered_map<int, std::vector<int>>& seedToTracksterAssociation) override;
+
+    void energyRegressionAndID(const std::vector<reco::CaloCluster>& layerClusters, std::vector<Trackster>& result);
+
+    static void fillPSetDescription(edm::ParameterSetDescription& iDesc);
+
+  private:
+    edm::ESGetToken<CaloGeometry, CaloGeometryRecord> caloGeomToken_;
```
Comment (perrotta) [inline RecoHGCal/TICL/plugins/PatternRecognitionbyFastJet.h:32]: This should also be `const` (but it is not fundamental, and I won't reset so many signatures only for that...)

**Comment by perrotta** [comment]
+1

**Pull request merged**